plugin/trino-hive/src/main/java/io/trino/plugin/hive/avro/AvroHiveFileWriterFactory.java

```diff
@@ -0,0 +1,143 @@
+import static io.trino.plugin.hive.util.HiveClassNames.AVRO_CONTAINER_OUTPUT_FORMAT_CLASS;
+import static io.trino.plugin.hive.util.HiveUtil.getColumnNames;
+import static io.trino.plugin.hive.util.HiveUtil.getColumnTypes;
+import static java.util.Objects.requireNonNull;
+
+public class AvroHiveFileWriterFactory
+        implements HiveFileWriterFactory
+{
+    private final TrinoFileSystemFactory fileSystemFactory;
+    private final TypeManager typeManager;
+    private final NodeVersion nodeVersion;
+
+    @Inject
+    public AvroHiveFileWriterFactory(
+            TrinoFileSystemFactory fileSystemFactory,
+            TypeManager typeManager,
+            NodeVersion nodeVersion)
+    {
+        this.fileSystemFactory = requireNonNull(fileSystemFactory, "hdfsEnvironment is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.nodeVersion = requireNonNull(nodeVersion, "nodeVersion");
+    }
+
+    @Override
+    public Optional<FileWriter> createFileWriter(
+            Location location,
+            List<String> inputColumnNames,
+            StorageFormat storageFormat,
+            HiveCompressionCodec compressionCodec,
+            Properties schema,
+            ConnectorSession session,
+            OptionalInt bucketNumber,
+            AcidTransaction transaction,
+            boolean useAcidSchema,
+            WriterKind writerKind)
+    {
+        if (!isAvroNativeWriterEnabled(session)) {
+            return Optional.empty();
+        }
+        if (!AVRO_CONTAINER_OUTPUT_FORMAT_CLASS.equals(storageFormat.getOutputFormat())) {
+            return Optional.empty();
+        }
+        checkArgument(!useAcidSchema, "Acid is not supported");
+
+        AvroCompressionKind compressionKind = compressionCodec.getAvroCompressionKind().orElse(AvroCompressionKind.NULL);
+        if (!compressionKind.isSupportedLocally()) {
+            throw new TrinoException(HIVE_WRITER_OPEN_ERROR, "Avro Compression codec %s is not supported in the environment".formatted(compressionKind));
+        }
+
+        HiveTimestampPrecision hiveTimestampPrecision = getTimestampPrecision(session);
+        // existing tables and partitions may have columns in a different order than the writer is providing, so build
+        // an index to rearrange columns in the proper order
+        List<String> fileColumnNames = getColumnNames(schema);
+        List<Type> fileColumnTypes = getColumnTypes(schema).stream()
+                .map(hiveType -> hiveType.getType(typeManager, hiveTimestampPrecision))
+                .collect(toImmutableList());
+
+        List<Type> inputColumnTypes = inputColumnNames.stream().map(inputColumnName -> {
+            int idx = fileColumnNames.indexOf(inputColumnName);
+            if (idx < 0) {
+                throw new TrinoException(HIVE_WRITER_OPEN_ERROR, "Input column name [%s] not preset in file columns names %s".formatted(inputColumnName, fileColumnNames));
```

Comment:
This should be `checkArgument()` since it is a problem for the caller (not a runtime error opening a file)